fairlearn/adversarial/_adversarial_mitigation.py

```diff
@@ -0,0 +1,822 @@
+            if select:
+                self.backend_ = PytorchEngine
+                return
+        if self.backend == "tensorflow" or self.backend == "auto":
+            select = False
+            try:
+                from tensorflow.keras import Model as model
+
+                tensorflow_installed = True
+                if isinstance(
+                    self.predictor_model, (list, model)
+                ) and isinstance(self.adversary_model, (list, model)):
+                    select = True
+                elif self.backend == "tensorflow":
+                    raise ValueError(
+                        _KWARG_ERROR_MESSAGE.format(
+                            "predictor_model and adversary_model",
+                            "a list or tensorflow.keras.Model",
+                        )
+                    )
+            except ImportError:
+                if self.backend == "tensorflow":
+                    raise RuntimeError(
+                        _IMPORT_ERROR_MESSAGE.format("tensorflow")
+                    )
+            if select:
+                self.backend_ = TensorflowEngine
+                return
+        # The keyword self.backend was weird
+        if self.backend not in ["torch", "tensorflow", "auto"]:
+            raise ValueError(
+                _KWARG_ERROR_MESSAGE.format(
+                    "backend", "one of ['auto', 'torch','tensorflow']"
+                )
+            )
+        # Or no backend is installed
+        if not (torch_installed or tensorflow_installed):
+            raise RuntimeError(
+                _IMPORT_ERROR_MESSAGE.format("torch or tensorflow")
+            )
+        # Or all other cases, a mismatch between model model (and installation)
+        raise ValueError(
+            _KWARG_ERROR_MESSAGE.format(
+                "predictor_model and adversary_model",
+                "a list, torch.nn.Module, or tensorflow.keras.Model. Also, "
+                + "make sure to have installed the corresponding backend",
+            )
+        )
+
+    def _set_predictor_function(self):
+        """
+        Infer prediction function.
+
+        The :code:`predictor_function` should be either the string
+        :code:`'auto'`, :code:`'binary'`, :code:`'category'`,
+        :code:`'continuous'`, or a callable. The string
+        :code:`'auto'` indicates to infer the predictor function
+        from the distribution type of :code:`y`
+        A callable should be a
+        function that maps the continuous output of the predictor model to
+        a discrete prediction.
+        """
+        if callable(self.predictor_function_):
+            pass
+        elif isinstance(self.predictor_function_, str):
+            kw = self.predictor_function_
+            if kw == "binary":
+                self.predictor_function_ = lambda pred: (pred >= 0.5).astype(
+                    float
+                )
+            elif kw == "category":
+
+                def loss(pred):
```

Comment:
Whence is this coming?

Reply:
I'm not sure what you mean?

Reply:
Can you explain the origin of this loss function? Why is it defined deeply embedded here, rather than somewhere it can be unit tested?

Reply:
@riedgar-ms sorry, i missed this message somehow. It seems trivial, but I would understand if you'd like this extracted and unit tested